### ISSUE TITLE
gdl crash because it does not find wxwidgets driver

### ISSUE THREAD

**Issue by brandy125** [gnudatalanguage/gdl#1050]
Just configured a fresh GDL download with

cmake .. -DHDF=OFF -DPYTHON=OFF

OpenMP support         ON (flag: -fopenmp)
WxWidgets              ON (libs:-L/usr/lib64;-pthread;;;-lwx_baseu-3.1;-lwx_gtk2u_core-3.1; headers:/usr/lib64/wx/include/gtk2-unicode-3.1;/usr/include/wx-3.1)
GRAPHICSMAGICK         ON (libs:/usr/lib64/libGraphicsMagick.so;/usr/lib64/libGraphicsMagick++.so; headers:/usr/include/GraphicsMagick)
TIFF                   ON (libs:/usr/lib64/libtiff.so; headers:/usr/include)
GeoTIFF                ON (libs:/usr/lib64/libgeotiff.so; headers:/usr/include/libgeotiff)
NetCDF                 ON (libs:netcdf; headers:)
HDF4                   OFF
HDF5                   ON (libs:/usr/lib64/libhdf5.so;/usr/lib64/libpthread.so;/usr/lib64/libz.so;/usr/lib64/libdl.so;/usr/lib64/libm.so; headers:/usr/include)
FFTW                   ON (libs:/usr/lib64/libfftw3.so;/usr/lib64/libfftw3f.so; headers:/usr/include)
MPI                    OFF
PROJ                   ON (libs:/usr/lib64/libproj.so; headers:/usr/include)
Python                 OFF
UDUNITS-2              ON (libs:/usr/lib64/libudunits2.so; headers:/usr/include)
EIGEN3                 ON (libs:; headers:/usr/include/eigen3)
GRIB                   ON (libs:/usr/lib64/libeccodes.so; headers:/usr/include)
GLPK                   ON (libs:/usr/lib64/libglpk.so; headers:/usr/include)
SHAPELIB               ON (libs:/usr/lib64/libshp.so; headers:/usr/include)
EXPAT                  ON (libs:/usr/lib64/libexpat.so; headers:/usr/include)
Xlib                   OFF
libpng                 ON (libs:/usr/lib64/libpng.so;/usr/lib64/libz.so; headers:/usr/include/libpng12;/usr/include)

-- Mandatory modules
Plplot                 ON (libs:/usr/local/lib/libplplot.so;/usr/local/lib/libplplotcxx.so; headers:/usr/local/include)
GNU Readline           ON (libs:/usr/lib64/libreadline.so;/usr/lib64/libhistory.so; headers:/usr/include)
GSL                    ON (libs:/usr/lib64/libgsl.so;/usr/lib64/libgslcblas.so; headers:/usr/include)
Zlib                   ON (libs:/usr/lib64/libz.so; headers:/usr/include)
(N)curses              ON (libs:/usr/lib64/libncurses.so;/usr/lib64/libform.so; headers:/usr/include)
RPC                    ON (libs:/usr/lib64/libtirpc.so; headers:/usr/include)

Python 3.8.10 is actually installed but GDL does not find it.
HDF5 is also installed but GDL does not find as well 

Compilation went fine

When I now start gdl and want to open a window then it crashes:

bash$ ./gdl
GDL - GNU Data Language, Version 1.0.0-rc.3 git
GDL> window,1
Unable to load driver: wxwidgets.

*** PLPLOT ERROR, IMMEDIATE EXIT ***
Unable to load driver
Program aborted

I read the README but did not understand why GDL is supposed to crash with my configuration.

Can anybody help?

My system is Linux Suse_Tumbleweed 5.13.2-1-default

**Comment by GillesDuvert** [comment]
you do not have the plplot wxwidget driver installed.
It is often packaged separately, probably under the name plplot-driver-wxwidgets

**Comment by GillesDuvert** [comment]
Or, alternatively, the PLPLOT driver location is unknown to the system. This because some plplot-related package has not been installed. Try install plplot, not the plplot-devel only.
Apparently Set the following environment variables:
PLPLOT_DRV_DIR to /usr/[lib?]/plplotx.y.z/drivers
PLPLOT_LIB to /usr/share/plplotx.y.z/
the latter directory holds the .pal and .fnt files

**Comment by brandy125** [comment]
Thank you for your remarks.

plplot and -dev was installed but the path was not exported to the shell environment when executing the gdl program.

After setting the environment variables correctly the GDL does not crash any more.

Even if it should be clear that the environment variables should be set correctly, there was no hint form GDL during configuration/compilation.
This may cause many users to have the same problem and make them desist from using GDL if it crashes with this message:

GDL> window,1
Unable to load driver: wxwidgets.

*** PLPLOT ERROR, IMMEDIATE EXIT ***
Unable to load driver
Program aborted
bash$ 

What do you think?

**Comment by GillesDuvert** [comment]
It is a worry indeed.
When plplot is built, the plpot 'suite' knows where it is installed. Normally, all the components share the same build location, so for a packaged GDL distribution I cannot imagine this can happen.
So, what causes the present behaviour? A plplot-devel not in line with the libplot?

**Comment by olebole** [comment]
I am currently packaging GDL for Debian, and I explicitly include the wxwindows driver in the dependencies of the packages. And this part seems to work.

**Comment by jeget** [comment]
I have a problem with WxWidgets of another kind...
I compiled GDL 1.0.1 and ran `./scripts/build_gdl.sh check` which finished successfully. When starting GDL, I see a message 

> -Using WxWidgets as graphics library (windows and widgets).

But if I type `window,1` I get an error

> ASSERT INFO:
> ../src/common/object.cpp(251): assert "classTable->Get(m_className) == NULL" failed in Register(): Class "wxCommandEvent" already in RTTI table - have you used IMPLEMENT_DYNAMIC_CLASS() multiple times or linked some object file twice)?
> BACKTRACE:
> [1] wxClassInfo::Register()
> [2] _dl_catch_exception
> [3] _dl_catch_exception
> [4] _dl_catch_error
> [5] dlopen
> [6] lt_dlopenadvise
> [7] lt_dlopenext
> [8] pllib_devinit
> [9] c_plinit
> [10] plstream::init()
> [11] GDLWXStream::GDLWXStream(int, int)
> [12] DeviceWX::WOpen(int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, int, int, int, bool)
> [13] lib::window(EnvT*)
> [14] PCALL_LIBNode::Run()
> [15] GDLInterpreter::statement(ProgNode*)
> [16] GDLInterpreter::call_fun(ProgNode*)
> [17] FCALLNode::Eval()
> [18] ARRAYEXPR_FCALLNode::Eval()
> [19] ASSIGN_REPLACENode::Run()
> [20] GDLInterpreter::statement(ProgNode*)
> [21] GDLInterpreter::call_pro(ProgNode*)
> [22] PCALLNode::Run()
> [23] GDLInterpreter::statement(ProgNode*)
> [24] GDLInterpreter::call_pro(ProgNode*)
> [25] PCALLNode::Run()
> [26] GDLInterpreter::statement(ProgNode*)
> [27] GDLInterpreter::call_pro(ProgNode*)
> [28] PCALLNode::Run()
> [29] GDLInterpreter::statement(ProgNode*)
> [30] GDLInterpreter::interactive(ProgNode*)
> [31] DInterpreter::ExecuteLine(std::istream*, unsigned long long)
> [32] DInterpreter::InterpreterLoop(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
> [33] main
> [34] __libc_start_main
> [35] _start

When configuring GDL, I first got a warning

>Cannot generate a safe runtime search path for target gdl because files in some directories may conflict with libraries in implicit directories

which was because of the libraries in anaconda and my build crashed. Then I removed anaconda from my $PATH, and configure and build finished with no warnings.
My system is Ubuntu 18.04.6 LTS.

Can anyone help?

**Comment by GillesDuvert** [comment]
@jeget I'm afraid the problem comes with GDL compiled with some version of wxWidgets while the driver in the plplot library is linked with another version of wxWidgets.  According to https://packages.ubuntu.com/ there are a number of wxWidgets libraries, an even greater number of plplot libraries an dependencies, I would expect some combination will work...

**Comment by jeget** [comment]
@GillesDuvert thanks for your reply! I looked up my installed packages. This is what `dpkg -l | grep wxwidgets` shows:

>ii  libplplotwxwidgets1:amd64                     5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (wxWidgets bindings)
>ii  plplot-driver-wxwidgets:amd64                 5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (wxWidgets driver)

and for plplot:

>ii  libplplot-dev                                 5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (development files)
>ii  libplplot-lua:amd64                           5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (Lua bindings)
>ii  libplplot-ocaml                               5.13.0+dfsg-6ubuntu2                            amd64        OCaml bindings for the scientific plotting library
>ii  libplplot15:amd64                             5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library
>ii  libplplotcxx13:amd64                          5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (C++ binding)
>ii  libplplotfortran0:amd64                       5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (fortran bindings)
>ii  libplplotqt2:amd64                            5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (Qt bindings)
>ii  libplplotwxwidgets1:amd64                     5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (wxWidgets bindings)
>ii  plplot-driver-qt                              5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (Qt driver)
>ii  plplot-driver-wxwidgets:amd64                 5.13.0+dfsg-6ubuntu2                            amd64        Scientific plotting library (wxWidgets driver)

Can I find out the wxwidgets and plplot versions GDL compiles with?
Also, in gdl/README I found the following piece:

>  The wxWidget version must be 2.8.1 or higher. 
>  WIDGET_DRAW 'plot' widgets use the plplot wxwidgets driver behind the scene. 
>  This driver must have been installed on your machine, it is not necessarily part of the base plplot distribution.
>  Unfortunately the wxWidget implementation of the plplot graphic library has changed between plplot 5.9.9 and
>  plplot 5.11, and several distributions have been affected by these incompatibilities. Should be solved now.

Does it mean I should downgrade plplot to, preferably, 5.9?

**Comment by GillesDuvert** [comment]
@jeget  It occurs to me that the plplot 'wxwidgets driver' (which is installed on your system, as you see the dpkg result:
`plplot-driver-wxwidgets:amd64 5.13.0+dfsg-6ubuntu2 amd64 Scientific plotting library (wxWidgets driver)` )
is somehow not found when you start GDL.
The problem may be solved by assigning the location of where the driver is (`wxwidgets.so` , located in something like `/usr/lib/x86_64-linux-gnu/plplot5.15.0/drivers`) to the environment variable ` PLPLOT_DRV_DIR` like in:
`$ export PLPLOT_DRV_DIR="/usr/lib/x86_64-linux-gnu/plplot5.15.0/drivers"`

**Comment by jeget** [comment]
@GillesDuvert I did `export PLPLOT_DRV_DIR="/usr/lib/x86_64-linux-gnu/plplot5.13.0/drivers"` which is where `wxwidgets.so` resides. Also I did `export PLPLOT_LIB="/usr/share/plplot5.13.0"` as suggested in the `INSTALL.plplot` text file. The error did not disappear.
However, I installed the `libwxgtk3.0-dev` package (previously I had `libwxgtk3.0-gtk3-dev`) and `window,1` works now. The difference I noticed is that previously, `./scripts/build_gdl.sh configure` wrote

>-- Found wxWidgets: -L/usr/lib/x86_64-linux-gnu;-pthread;;;-lwx_baseu-3.0;-lwx_gtk3u_core-3.0;-lwx_gtk3u_adv-3.0 (found version "3.0.4") 

and now

>-- Found wxWidgets: -L/usr/lib/x86_64-linux-gnu;-pthread;;;-lwx_baseu-3.0;-lwx_gtk2u_core-3.0;-lwx_gtk2u_adv-3.0 (found version "3.0.4")

A side effect is that now I get a warning

>Gtk-Message: 02:59:51.257: Failed to load module "canberra-gtk-module"

which was not there earlier. Hope this will not cause problems...
Thanks for your help!
